File: packages/metadata/python/metadata/loader.py
"""YAML loader that builds a typed Registry from data/."""
from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import yaml

from .schema import (
    ConnectorPattern,
    KpiMasterEntry,
    KpiRegistryEntry,
    KpiSqlSpec,
    SourceSystemEntry,
    Subdomain,
)


def _read_yaml_dir(d: Path, *, exclude: tuple[str, ...] = ()) -> list[Any]:
    if not d.exists():
        return []
    out: list[Any] = []
    for f in sorted(d.iterdir()):
        if f.suffix.lower() not in {".yaml", ".yml"}:
            continue
        if f.name in exclude:
            continue
        with f.open("r", encoding="utf-8") as fh:
            out.append(yaml.safe_load(fh))
    return out


def _read_yaml(file: Path) -> Any:
    if not file.exists():
        return None
    with file.open("r", encoding="utf-8") as fh:
        return yaml.safe_load(fh)


@dataclass
class Registry:
    subdomains: list[Subdomain]
    kpis: list[KpiRegistryEntry]
    source_systems: list[SourceSystemEntry]
    connectors: list[ConnectorPattern]
    kpi_master: list[KpiMasterEntry] = field(default_factory=list)
    kpi_sql: list[KpiSqlSpec] = field(default_factory=list)


def load_registry(data_root: Path | str = "data") -> Registry:
    root = Path(data_root)
    subdomains = [
        Subdomain.model_validate(raw) for raw in _read_yaml_dir(root / "taxonomy")
    ]
    kpis: list[KpiRegistryEntry] = []
    # Skip master.yaml + sql.yaml here; they're loaded via dedicated helpers
    # because they carry richer (forbidden-extra) fields.
    for raw in _read_yaml_dir(root / "kpis", exclude=("master.yaml", "sql.yaml")):
        for k in (raw or {}).get("kpis", []):
            try:
                kpis.append(KpiRegistryEntry.model_validate(k))
            except Exception:
                # Tolerate stray entries that don't fit the registry shape
                # (e.g. master entries without `vertical`).
                continue
    sources: list[SourceSystemEntry] = []
    for raw in _read_yaml_dir(root / "source-systems"):
        for s in (raw or {}).get("sources", []):
            sources.append(SourceSystemEntry.model_validate(s))
    connectors: list[ConnectorPattern] = []
    for raw in _read_yaml_dir(root / "connectors"):
        for c in (raw or {}).get("connectors", []):
            connectors.append(ConnectorPattern.model_validate(c))
    kpi_master = load_kpi_master(root)
    kpi_sql = load_kpi_sql(root)
    return Registry(
        subdomains=subdomains,
        kpis=kpis,
        source_systems=sources,
        connectors=connectors,
        kpi_master=kpi_master,
        kpi_sql=kpi_sql,
    )


def load_kpi_master(data_root: Path | str = "data") -> list[KpiMasterEntry]:
    """Load `data/kpis/master.yaml` if it exists, otherwise return []."""
    root = Path(data_root)
    raw = _read_yaml(root / "kpis" / "master.yaml")
    if not raw:
        return []
    return [KpiMasterEntry.model_validate(k) for k in (raw.get("kpis") or [])]


def load_kpi_sql(data_root: Path | str = "data") -> list[KpiSqlSpec]:
    """Load `data/kpis/sql.yaml` if it exists, otherwise return []."""
    root = Path(data_root)
    raw = _read_yaml(root / "kpis" / "sql.yaml")
    if not raw:
        return []
    return [KpiSqlSpec.model_validate(k) for k in (raw.get("kpis") or [])]


def get_subdomain(registry: Registry, sub_id: str) -> Subdomain | None:
    return next((s for s in registry.subdomains if s.id == sub_id), None)


def get_kpi(registry: Registry, kpi_id: str) -> KpiRegistryEntry | None:
    return next((k for k in registry.kpis if k.id == kpi_id), None)


def get_source_system(registry: Registry, source_id: str) -> SourceSystemEntry | None:
    return next((s for s in registry.source_systems if s.id == source_id), None)

File: packages/metadata/src/loader.ts
import { existsSync, readFileSync, readdirSync } from "node:fs";
import { join, resolve } from "node:path";
import { parse as parseYaml } from "yaml";
import {
  ConnectorPattern,
  GlossaryTerm,
  KpiMasterEntry,
  KpiRegistryEntry,
  KpiSqlSpec,
  SourceSystemEntry,
  Subdomain,
} from "./schema";
import type {
  ConnectorPattern as ConnectorPatternT,
  GlossaryTerm as GlossaryTermT,
  KpiMasterEntry as KpiMasterEntryT,
  KpiRegistryEntry as KpiRegistryEntryT,
  KpiSqlSpec as KpiSqlSpecT,
  SourceSystemEntry as SourceSystemEntryT,
  Subdomain as SubdomainT,
} from "./schema";

const DATA_ROOT = resolve(process.cwd(), "data");

function readYamlDir(dir: string, exclude: string[] = []): unknown[] {
  let entries: string[];
  try {
    entries = readdirSync(dir);
  } catch {
    return [];
  }
  return entries
    .filter((f) => f.endsWith(".yaml") || f.endsWith(".yml"))
    .filter((f) => !exclude.includes(f))
    .map((f) => parseYaml(readFileSync(join(dir, f), "utf8")));
}

function readYamlFile(file: string): unknown {
  if (!existsSync(file)) return null;
  return parseYaml(readFileSync(file, "utf8"));
}

export interface Registry {
  subdomains: SubdomainT[];
  kpis: KpiRegistryEntryT[];
  sourceSystems: SourceSystemEntryT[];
  connectors: ConnectorPatternT[];
  glossary: GlossaryTermT[];
  kpiMaster: KpiMasterEntryT[];
  kpiSql: KpiSqlSpecT[];
}

export function loadRegistry(dataRoot: string = DATA_ROOT): Registry {
  const subdomains = readYamlDir(join(dataRoot, "taxonomy")).map((raw) =>
    Subdomain.parse(raw),
  );
  // Skip master.yaml + sql.yaml here; they're loaded via dedicated helpers
  // because they carry richer (forbidden-extra) fields and may not have
  // `vertical` populated for every entry.
  const kpis = readYamlDir(join(dataRoot, "kpis"), ["master.yaml", "sql.yaml"])
    .flatMap((raw) => {
      const list = (raw as { kpis?: unknown[] } | null)?.kpis ?? [];
      return list;
    })
    .map((k) => {
      try {
        return KpiRegistryEntry.parse(k);
      } catch {
        return null;
      }
    })
    .filter((k): k is KpiRegistryEntryT => k !== null);
  const sourceSystems = readYamlDir(join(dataRoot, "source-systems")).flatMap(
    (raw) => {
      const list = (raw as { sources?: unknown[] } | null)?.sources ?? [];
      return list.map((s) => SourceSystemEntry.parse(s));
    },
  );
  const connectors = readYamlDir(join(dataRoot, "connectors")).flatMap(
    (raw) => {
      const list = (raw as { connectors?: unknown[] } | null)?.connectors ?? [];
      return list.map((c) => ConnectorPattern.parse(c));
    },
  );
  const glossary = readYamlDir(join(dataRoot, "glossary")).flatMap((raw) => {
    const list = (raw as { terms?: unknown[] } | null)?.terms ?? [];
    return list.map((t) => GlossaryTerm.parse(t));
  });
  const masterRaw = readYamlFile(join(dataRoot, "kpis", "master.yaml")) as
    | { kpis?: unknown[] }
    | null;
  const kpiMaster = (masterRaw?.kpis ?? []).map((k) => KpiMasterEntry.parse(k));
  const sqlRaw = readYamlFile(join(dataRoot, "kpis", "sql.yaml")) as
    | { kpis?: unknown[] }
    | null;
  const kpiSql = (sqlRaw?.kpis ?? []).map((k) => KpiSqlSpec.parse(k));
  return {
    subdomains,
    kpis,
    sourceSystems,
    connectors,
    glossary,
    kpiMaster,
    kpiSql,
  };
}

export function getSubdomain(
  registry: Registry,
  id: string,
): SubdomainT | undefined {
  return registry.subdomains.find((s) => s.id === id);
}

export function getSubdomainsByVertical(
  registry: Registry,
  vertical: string,
): SubdomainT[] {
  return registry.subdomains.filter((s) => s.vertical === vertical);
}

export function getKpi(
  registry: Registry,
  id: string,
): KpiRegistryEntryT | undefined {
  return registry.kpis.find((k) => k.id === id);
}

export function getSourceSystem(
  registry: Registry,
  id: string,
): SourceSystemEntryT | undefined {
  return registry.sourceSystems.find((s) => s.id === id);
}

export function getKpiMaster(
  registry: Registry,
  id: string,
): KpiMasterEntryT | undefined {
  return registry.kpiMaster.find((k) => k.id === id);
}

export function getKpiSql(
  registry: Registry,
  id: string,
): KpiSqlSpecT | undefined {
  return registry.kpiSql.find((k) => k.kpi_id === id);
}
